### PR TITLE
Fix/wpds no oop languages

### PR DESCRIPTION
**Pull request by JulianSchuette**
This PR improves precision of typestate analysis to better handle initial configurations and non-OO languages

### CODE REVIEW

**Comment by oxisto** [comment]
I've resolved the merge conflicts and updated it to the current master.

File: src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java [hunk 1]
```diff
@@ -267,6 +267,23 @@ public static String toNonQualifiedName(@NonNull String typeStr) {
...
 	}
 
+	/**
+	 * Counterpart to toNonQualifiedName().
+	 *
+	 * @param fqn
+	 * @return
+	 */
+	@NonNull
+	public static String getScope(@NonNull String fqn) {
```
Comment (fwendland) [inline src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java:276]: Can you explain, what this method intends to achieve? Are you trying to get the namespace? What if namespaces are nested?

File: src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java [hunk 2]
```diff
@@ -267,6 +267,23 @@ public static String toNonQualifiedName(@NonNull String typeStr) {
...
+		int posDot = fqn.lastIndexOf('.');
+		int posColon = fqn.indexOf(':');
```
Comment (fwendland) [inline src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java:278]: Shouldn't this be `fqn.lastIndexOf(':') - 1`?

**Comment by fwendland** [comment]
> 
> 
> I've resolved the merge conflicts and updated it to the current master.

Thank you for the save. :)

File: src/main/java/de/fraunhofer/aisec/analysis/wpds/TypeStateAnalysis.java [hunk 3]
```diff
@@ -883,19 +948,22 @@ private boolean isPhantom(CallExpression mce) {
...
-
 		// Return site(s). Actually, multiple return sites will only occur in case of exception handling.
 		List<Vertex> returnSites = CrymlinQueryWrapper.getNextStatements(crymlinTraversal, (long) currentStmtVertex.id());
 
 		// Arguments of function call
 		List<Val> argVals = argumentsToVals(mce, currentFunctionName);
 
+		Set<NFATransition<Node>> relevantNFATransitions = new HashSet<>();
+		for (Val argVal : argVals) {
+			nfa.getTransitions()
```
[the remark this answers is not in the document]
Comment (JulianSchuette) [reply src/main/java/de/fraunhofer/aisec/analysis/wpds/TypeStateAnalysis.java:926]: This code is a relict. A push rule will not trigger a typestate transition, so there is nothing to be filled here. I'll remove it.

File: src/main/java/de/fraunhofer/aisec/analysis/wpds/TypeStateAnalysis.java [hunk 4]
```diff
@@ -883,19 +948,22 @@ private boolean isPhantom(CallExpression mce) {
...
+					.stream()
+					.filter(
+						tran -> belongsToOp(argVal, mce, tran.getTarget().getBase(), tran.getTarget().getOp()))
+					.collect(Collectors.toSet());
+		}
+		TypestateWeight weight = relevantNFATransitions.isEmpty() ? TypestateWeight.one() : new TypestateWeight(relevantNFATransitions);
```
[the remark this answers is not in the document]
Comment (JulianSchuette) [reply src/main/java/de/fraunhofer/aisec/analysis/wpds/TypeStateAnalysis.java:932]: Yes, see above.

File: src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java (hunk 1 above)
Comment (JulianSchuette) [reply src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java:276]: Yes, this method should get "everything before the actual name of the type", i.e. a namespace in C++ and a package in Java.

`java.lang.String` -> `java.lang`
`Botan::Cipher_Mode::create` -> `Botan::Cipher_Mode`

The only purpose of this method is currently to support in the comparison of `type`s as provided by the CPG with fully qualified names of method calls in MARK. For instance, this is how it looks for Java:

```
op bla {
  java.lang.String.toString()
}
```

If the code contains a line `x.toString()`, Codyze gets the `type` of base `x` and compares it against the type given in  MARK. 
This is how it looks for C++:

```
op bla {
  Botan::Cipher_Mode::create(_,_,_,_);
}
```

File: src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java (hunk 2 above)
Comment (JulianSchuette) [reply src/main/java/de/fraunhofer/aisec/analysis/utils/Utils.java]: Almost, yes. It should be `lastIndexOf("::")`. I added the fix to this PR.

**Review by fwendland — APPROVED** [review]

**Pull request merged**